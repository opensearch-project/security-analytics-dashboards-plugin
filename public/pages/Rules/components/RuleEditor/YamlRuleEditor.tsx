/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { dump, load } from 'js-yaml';
import { EuiFormRow, EuiCodeEditor, EuiLink, EuiSpacer, EuiText, EuiForm } from '@elastic/eui';
import FormFieldHeader from '../../../../components/FormFieldHeader';
import { Rule } from '../../../../../models/interfaces';
import {
  AUTHOR_REGEX,
  validateDescription,
  validateName,
  authorErrorString,
  descriptionErrorString,
  titleErrorString,
} from '../../../../utils/validation';

export interface YamlRuleEditorProps {
  rule: Rule;
  change: React.Dispatch<Rule>;
}

export interface YamlEditorState {
  errors: string[] | null;
  value?: string;
}

const mapYamlObjectToYamlString = (rule: Rule): string => {
  try {
    if (!rule.detection) {
      const { detection, ...ruleWithoutDetection } = rule;
      return dump(ruleWithoutDetection);
    } else {
      return dump(rule);
    }
  } catch (error: any) {
    console.warn('Security Analytics - Rule Eritor - Yaml dump', error);
    return '';
  }
};

const mapRuleToYamlObject = (rule: Rule): any => {
  let detection = undefined;
  if (rule.detection) {
    try {
      detection = load(rule.detection);
    } catch {}
  }

  const yamlObject: any = {
    id: rule.id,
    logsource: { product: rule.category },
    title: rule.title,
    description: rule.description,
    tags: rule.tags.map((tag) => tag.value),
    falsepositives: rule.false_positives.map((falsePositive) => falsePositive.value),
    level: rule.level,
    status: rule.status,
    references: rule.references.map((reference) => reference.value),
    author: rule.author,
    detection,
  };

  return yamlObject;
};

const mapYamlObjectToRule = (obj: any): Rule => {
  let detection = '';
  if (obj.detection) {
    try {
      detection = dump(obj.detection);
    } catch {}
  }
  const rule: Rule = {
    id: obj.id,
    category: obj.logsource ? obj.logsource.product : undefined,
    log_source: '',
    title: obj.title,
    description: obj.description,
    tags: obj.tags ? obj.tags.map((tag: string) => ({ value: tag })) : undefined,
    false_positives: obj.falsepositives
      ? obj.falsepositives.map((falsePositive: string) => ({ value: falsePositive }))
      : undefined,
    level: obj.level,
    status: obj.status,
    references: obj.references
      ? obj.references.map((reference: string) => ({ value: reference }))
      : undefined,
    author: obj.author,
    detection,
  };

  return rule;
};

const validateRule = (rule: Rule): string[] | null => {
  const requiredFiledsValidationErrors: Array<string> = [];

  if (!rule.title) {
    requiredFiledsValidationErrors.push('Title is required');
  }
  if (!rule.category) {
    requiredFiledsValidationErrors.push('Logsource is required');
  }
  if (!rule.level) {
    requiredFiledsValidationErrors.push('Level is required');
  }
  if (!rule.author) {
    requiredFiledsValidationErrors.push('Author is required');
  }
  if (!rule.status) {
    requiredFiledsValidationErrors.push('Status is required');
  }

  if (requiredFiledsValidationErrors.length > 0) {
    return requiredFiledsValidationErrors;
  }

  if (!validateName(rule.title, AUTHOR_REGEX)) {
    return [titleErrorString];
  }
  if (!validateDescription(rule.description)) {
    return [descriptionErrorString];
  }
  if (!validateName(rule.author, AUTHOR_REGEX)) {
    return [authorErrorString];
  }

  return null;
};

export const YamlRuleEditor: React.FC<YamlRuleEditorProps> = ({ rule, change }) => {
  const yamlObject = mapRuleToYamlObject(rule);

  const [state, setState] = useState<YamlEditorState>({
    errors: null,
    value: mapYamlObjectToYamlString(yamlObject),
  });

  const onChange = (value: string) => {
    setState((prevState) => ({ ...prevState, value }));
  };

  const onBlur = () => {
    if (!state.value) {
      setState((prevState) => ({ ...prevState, errors: ['Rule cannot be empty'] }));
      return;
    }
    try {
      const yamlObject = load(state.value);

      const rule = mapYamlObjectToRule(yamlObject);

      const errors = validateRule(rule);

      if (errors && errors.length > 0) {
        setState((prevState) => ({ ...prevState, errors: errors }));
        return;
      }

      change(rule);
      setState((prevState) => ({ ...prevState, errors: null }));
    } catch (error) {
      setState((prevState) => ({ ...prevState, errors: ['Invalid YAML'] }));

      console.warn('Security Analytics - Rule Eritor - Yaml load', error);
    }
  };

  return (
    <>
      <EuiForm
        isInvalid={state.errors !== null && state.errors.length > 0}
        error={state.errors}
        component="form"
      >
        <EuiFormRow
          label={<FormFieldHeader headerTitle={'Define rule in YAML'} />}
          fullWidth={true}
        >
          <>
            <EuiText size="s" color="subdued">
              Use the YAML editor to define a sigma rule. See{' '}
              <EuiLink href="https://github.com/SigmaHQ/sigma-specification">
                Sigma specification
              </EuiLink>{' '}
              for rule structure and schema.
            </EuiText>
            <EuiSpacer size="s" />
            <EuiCodeEditor
              mode="yaml"
              width="100%"
              value={state.value}
              onChange={onChange}
              onBlur={onBlur}
              data-test-subj={'rule_yaml_editor'}
            />
          </>
        </EuiFormRow>
      </EuiForm>
    </>
  );
};
