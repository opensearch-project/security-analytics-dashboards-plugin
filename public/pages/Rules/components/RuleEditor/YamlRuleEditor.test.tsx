/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from '@testing-library/react';
import { YamlRuleEditor } from './YamlRuleEditor';

describe('<YamlRuleEditor /> spec', () => {
  it('renders the component', () => {
    const { container } = render(
      <YamlRuleEditor
        change={() => {}}
        rule={{
          id: '25b9c01c-350d-4b95-bed1-836d04a4f324',
          category: 'windows',
          title: 'Testing rule',
          description: 'Testing Description',
          status: 'experimental',
          author: 'Bhabesh Raj',
          references: [
            {
              value: 'https://securelist.com/operation-tunnelsnake-and-moriya-rootkit/101831',
            },
          ],
          tags: [
            {
              value: 'attack.persistence',
            },
            {
              value: 'attack.privilege_escalation',
            },
            {
              value: 'attack.t1543.003',
            },
          ],
          log_source: '',
          detection:
            'selection:\n  Provider_Name: Service Control Manager\n  EventID: 7045\n  ServiceName: ZzNetSvc\ncondition: selection\n',
          level: 'high',
          false_positives: [
            {
              value: 'Unknown',
            },
          ],
        }}
      >
        <div>Testing YamlRuleEditor</div>
      </YamlRuleEditor>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
