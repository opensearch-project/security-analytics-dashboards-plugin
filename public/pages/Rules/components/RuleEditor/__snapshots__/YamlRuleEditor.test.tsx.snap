// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YamlRuleEditor /> spec renders the component 1`] = `
<form
  class="euiForm"
>
  <div
    class="euiFormRow euiFormRow--fullWidth"
    id="some_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="some_html_id"
      >
        <div
          class="euiText euiText--small"
        >
          <strong>
            Define rule in YAML
          </strong>
        </div>
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiText euiText--small"
      >
        <div
          class="euiTextColor euiTextColor--subdued"
        >
          Use the YAML editor to define a sigma rule. See
           
          <a
            class="euiLink euiLink--primary"
            href="https://github.com/SigmaHQ/sigma-specification"
            rel="noreferrer"
          >
            Sigma specification
          </a>
           
          for rule structure and schema.
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div
        class="euiCodeEditorWrapper"
        data-test-subj="rule_yaml_editor"
        style="width: 100%;"
      >
        <button
          class="euiCodeEditorKeyboardHint"
          data-test-subj="codeEditorHint"
          id="some_html_id"
        >
          <p
            class="euiText"
          >
            Press Enter to start editing.
          </p>
          <p
            class="euiText"
          >
            When you're done, press Escape to stop editing.
          </p>
        </button>
        <div
          class=" ace_editor ace-tm"
          id="some_html_id"
          style="width: 100%; height: 500px;"
        >
          <textarea
            autocapitalize="off"
            autocorrect="off"
            class="ace_text-input"
            spellcheck="false"
            style="opacity: 0;"
            tabindex="-1"
            wrap="off"
          />
          <div
            aria-hidden="true"
            class="ace_gutter"
          >
            <div
              class="ace_layer ace_gutter-layer ace_folding-enabled"
            />
            <div
              class="ace_gutter-active-line"
            />
          </div>
          <div
            class="ace_scroller"
          >
            <div
              class="ace_content"
            >
              <div
                class="ace_layer ace_print-margin-layer"
              >
                <div
                  class="ace_print-margin"
                  style="left: 4px; visibility: visible;"
                />
              </div>
              <div
                class="ace_layer ace_marker-layer"
              />
              <div
                class="ace_layer ace_text-layer"
                style="padding: 0px 4px;"
              />
              <div
                class="ace_layer ace_marker-layer"
              />
              <div
                class="ace_layer ace_cursor-layer ace_hidden-cursors"
              >
                <div
                  class="ace_cursor"
                />
              </div>
            </div>
          </div>
          <div
            class="ace_scrollbar ace_scrollbar-v"
            style="display: none; width: 20px;"
          >
            <div
              class="ace_scrollbar-inner"
              style="width: 20px;"
            />
          </div>
          <div
            class="ace_scrollbar ace_scrollbar-h"
            style="display: none; height: 20px;"
          >
            <div
              class="ace_scrollbar-inner"
              style="height: 20px;"
            />
          </div>
          <div
            style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;"
          >
            <div
              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
            />
            <div
              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
            >
              XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
